### ISSUE TITLE
ANTs deformations in pytorch

### ISSUE THREAD

**Issue by A-Thorley** [ANTsX/ANTs#1205]
**The problem**

I have used ANTs to successfully register a pair of fixed and moving 3D Cardiac MRI images, applying the deformation produced by ants to a label corresponding to the moving image with AntsApplyTransform. A dice score of 0.83 of achieved between the warped moving label and fixed label. To produce some visualisations, I have read in the deformation to python with nibabel, scaled the displacement field to torchs [-1,1] system and used the torch grid sample method to warp the moving label. To check this was functioning correctly I checked the dice score using this approach and found it to be significantly lower than the 0.83 achieved using the AntsApplyTransform. I am unsure why this is happening, I understood the deformation by ants to be a displacement field in physical space. Am I misunderstanding the format of the ants deformation?

**To recreate**

The following is an example of how I am registering two images using Ants:

`"antsRegistration --dimensionality 3 --float 0 --output [SAVE_deformation,SAVE_warped,SAVE_Inversewarped] " \`
            `"--interpolation Linear " \`
            `"--winsorize-image-intensities [0.005,0.995] " \`
            `"--use-histogram-matching 0 " \`
            `"--transform SyN[{6},{7},{8}] " \`
            `"--metric CC[FIX,MOV,1,4] " \`
            `"--convergence [100x70x50x0,1e-6,10] " \`
            `"--shrink-factors 8x4x2x1 " \`
            `"--smoothing-sigmas 3x2x1x0vox --verbose 1 "`
			
			
`"antsApplyTransforms -d 3 " \`
                        `"-n NearestNeighbor " \`
                        `"-i {MOV} " \`
                        `"-r {MOV} " \`
                        `"-t {SAVE_warped} " \`
                        `"-o {WarpedMOV} --verbose 1 "`

The following is an example of how I am attempting to warp an image or label with the ants displacement field.

`import torch.nn.functional as f`
`import nibabel as nib`
`import torch`
`import numpy as np`

`image_path = "your path to image"`
`ants_displacements_path = "your path to displacements"`

`# Read in displacements`
`ants_disp_nii = nib.load(ants_displacements_path)`
`displacements = ants_disp_nii.get_fdata()`
`displacements = torch.from_numpy(displacements).squeeze().unsqueeze(0).unsqueeze(0).double()`

`# Read in image`
`image_nii = nib.load(image_path)`
`image = image_nii.get_fdata()`
`image = torch.from_numpy(image).squeeze().unsqueeze(0).unsqueeze(0).double()`

`# Split into x,y,z components`
`x_displacement = displacements[..., 0]`
`y_displacement = displacements[..., 1]`
`z_displacement = displacements[..., 2]`

`# Generate mesh ID grid`
`x, y, z = x_displacement.shape[-3:]`
`grid_x, grid_y, grid_z = torch.meshgrid(`
`    [torch.linspace(-1, 1, x, dtype=torch.double),`
`     torch.linspace(-1, 1, y, dtype=torch.double),`
`     torch.linspace(-1, 1, z, dtype=torch.double)])`

`# Normalise displacements to torch [-1,1] coord system and add ID grid`
`x_deformation = (grid_x + (x_displacement * 2 / x))`
`y_deformation = (grid_y + (y_displacement * 2 / y))`
`z_deformation = (grid_z + (z_displacement * 2 / z))`

`# Warp image with grid sample`
`deformation = torch.stack((deformation_z, deformation_y, deformation_x), 4)`
`warped_image = f.grid_sample(image, deformation, mode="nearest", padding_mode='zeros', align_corners=True)`

Any insight into where I might be going wrong with this would be greatly appreciated,

Alex

**Comment by ntustison** [comment]
One thing that needs to be corrected is ``-r {FIX}`` instead of ``-r {MOV}``:

```
"antsApplyTransforms -d 3 " \
"-n NearestNeighbor " \
"-i {MOV} " \
"-r {FIX} " \
"-t {SAVE_warped} " \
"-o {WarpedMOV} --verbose 1 "
```

**Comment by xi-jia** [comment]
> One thing that needs to be corrected is `-r {FIX}` instead of `-r {MOV}`:
> 
> ```
> "antsApplyTransforms -d 3 " \
> "-n NearestNeighbor " \
> "-i {MOV} " \
> "-r {FIX} " \
> "-t {SAVE_warped} " \
> "-o {WarpedMOV} --verbose 1 "
> ```
Yes, you are right.
But -r FIX or -r MOV doesn't matter,  as long as the FIX and MOV have the same (spacing, origin, size, and direction) in mono-modality registration.

**Comment by gdevenyi** [comment]
This is likely the relevant documentation on the underlying ITK implementation, https://itk.org/Doxygen/html/classitk_1_1DisplacementFieldTransform.html

**Comment by cookpa** [comment]
To apply the ANTs transform correctly, you need to do:

for each voxel in the output image (in fixed space):
  1. Transform voxel coordinate (i,j,k) to physical space (x,y,z)
  2. Apply forward warp field, moving point to (x',y',z')
  3. Get the appropriate interpolated value at (x',y',z') in the moving image
  4. Write the value to voxel (i,j,k) in the output image
done

With the inverse warps, you iterate over the moving image grid, and apply the inverse warp to get an interpolated value in the fixed space. This ordering has caused confusion before (https://github.com/ANTsX/ANTs/wiki/Forward-and-inverse-warps-for-warping-images,-pointsets-and-Jacobians).

ITK uses an "LPS" coordinate system, where

x increases from right towards Left 
y increases from anterior towards Posterior
z increases from inferior towards Superior

ITK coordinates are relative to the center of a voxel, not the corner. So index (0,0,0) should map to the center of the first voxel in physical space.

The displacement vectors are also written in the same coordinate system.

https://simpleitk.readthedocs.io/en/master/fundamentalConcepts.html

I hope this information is helpful.
